### PR TITLE
Fix for a crash in String.substring(with:) [SR-2483]

### PR DESCRIPTION
**Pull request by pushkarnk**
I think this is just one way to fix: https://bugs.swift.org/browse/SR-2483 
I've documented some additional findings in the bug report.

### CODE REVIEW

**Comment by phausler** [comment]
could you possibly add a unit test to verify this behavior?

**Comment by pushkarnk** [comment]
@phausler : I am no longer comfortable with this fix. The API reference says ```String.init?(_ utf16)``` returns nil only when the utf16 contains unpaired surrogates - so, the case of it returning nil for "\r\n"[1..<1] does not match the spec and we cannot simply return "". This needs more investigation on the stdlib part. If you don't mind, I'll keep this pull request open until we have a better understanding of how this should be fixed.

**Comment by phausler** [comment]
Certainly, it might be useful to verify the ground truth of how that behaves in objc - since that is our baseline here

**Comment by pushkarnk** [comment]
When using ```substring(with:)``` if either end of the range breaks a surrogate pair, we crash on Linux but not on Darwin. On Darwin, I found the broken surrogate pair being replaced by the[ Unicode replacement character](https://en.wikipedia.org/wiki/Specials_(Unicode_block)#Replacement_character). We need to do the same on Linux. Also, for ranges of the type ```t..<t``` we need to return "".

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
Updated this to include a handful of corner cases with surrogate pairs and `\r\n`

**Comment by pushkarnk** [comment]
@swift-ci please test

File: Foundation/NSString.swift [hunk 1]
```diff
@@ -346,7 +346,49 @@ extension NSString {
...
+
+            //if the range breaks a surrogate pair at the end of the string
+            if let substrPrefix = String(_storage.utf16[min..<max.advanced(by: -1)]) {
+                return substrPrefix + suffix
+            }
+
+            //the range probably breaks surrogate pairs at both the ends
+            guard min.advanced(by: 1) <= max.advanced(by: -1) else { return prefix + suffix }
+
+            let substr =  String(_storage.utf16[min.advanced(by: 1)..<max.advanced(by: -1)])!
```
Comment (pushkarnk) [inline Foundation/NSString.swift]: After all the prior checks, I think this can be safely unwrapped.

**Comment by pushkarnk** [comment]
@phausler @parkera Sorry for the delay on this. 

I've tried to make `substring(with: NSRange)` more robust (and behave like Darwin) here, in the process of fixing SR-2483. Can you please review? Thanks.

**Comment by parkera** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
@parkera @phausler is there anything else you'd want me to consider here?

**Comment by parkera** [comment]
This took me a while to review because I was digging through the Darwin version, which really defers this operation to CFString. That's where the replacement character substitution happens. In the long term we may want to make NSString.swift do something similar so we don't end up replicating this logic in two places, but in the short term this fix seems appropriate.

**Comment by parkera** [comment]
@swift-ci test and merge
